File: q5/q5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expandCenter(s, left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return right-left-1
    
        if (len(s) == 0):
            return ""
        end = 0
        start = 0
        for i in range(len(s)):
            len1 = expandCenter(s, i, i)
            len2 = expandCenter(s,i, i+1)
            length = max(len1,len2)
            if length > (end - start+1):
                if length % 2 == 1:
                    start = i - (length - 1)//2
                    end = i + (length - 1)//2
                else:
                    start = i - (length - 2)//2
                    end = i + length//2
        return s[start:end+1]

        
    

File: q5/q5.cpp

class Solution {
public:
    string longestPalindrome(string s) {
        if (s.size() == 0){
            return "";
        }
        if (s.size() == 1){
            return s;
        }
        int start = 0, end = 0;
        for (int i = 0; i < s.size(); i++){
            int len1 = expandCenter(s, i, i);
            int len2 = expandCenter(s, i, i+1);
            int len = len1 > len2 ? len1 : len2;
            if (len > end -start){
                start = i - (len - 1)/2;
                end = i + len / 2;
            }
        }
        return s.substr(start, end-start+1);
    }
    
    int expandCenter(string str, int left, int right){
        while(left >= 0 && right < str.size() && str[left] == str[right]){
            left --;
            right ++;
        }
        return right-left-1;
    }
};
